overlord/devicestate/devicestate_test.go

```diff
@@ -2771,6 +2793,95 @@ func (s *deviceMgrSuite) TestRemodelLessRequiredSnaps(c *C) {
 	c.Assert(tSetModel.Summary(), Equals, "Set new model assertion")
 }
 
+type freshSessionStore struct {
+	storetest.Store
+
+	ensureDeviceSession int
+}
+
+func (sto *freshSessionStore) EnsureDeviceSession() (*auth.DeviceState, error) {
+	sto.ensureDeviceSession += 1
+	return nil, nil
+}
+
+func (s *deviceMgrSuite) TestRemodelStoreSwitch(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+	s.state.Set("seeded", true)
+	s.state.Set("refresh-privacy-key", "some-privacy-key")
+
+	var testStore snapstate.StoreService
+
+	restore := devicestate.MockSnapstateInstallWithDeviceContext(func(st *state.State, name, channel, cohort string, revision snap.Revision, userID int, flags snapstate.Flags, deviceCtx snapstate.DeviceContext) (*state.TaskSet, error) {
+		c.Check(flags.Required, Equals, true)
+		c.Check(deviceCtx, NotNil)
+		c.Check(deviceCtx.ForRemodeling(), Equals, true)
+
+		c.Check(deviceCtx.Store(), Equals, testStore)
+
+		tDownload := s.state.NewTask("fake-download", fmt.Sprintf("Download %s", name))
+		tValidate := s.state.NewTask("validate-snap", fmt.Sprintf("Validate %s", name))
+		tValidate.WaitFor(tDownload)
+		tInstall := s.state.NewTask("fake-install", fmt.Sprintf("Install %s", name))
+		tInstall.WaitFor(tValidate)
+		ts := state.NewTaskSet(tDownload, tValidate, tInstall)
+		ts.MarkEdge(tValidate, snapstate.DownloadAndChecksDoneEdge)
+		return ts, nil
+	})
+	defer restore()
+
+	// set a model assertion
+	s.makeModelAssertionInState(c, "canonical", "pc-model", map[string]interface{}{
+		"architecture": "amd64",
+		"kernel":       "pc-kernel",
+		"gadget":       "pc",
+		"base":         "core18",
+	})
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand: "canonical",
+		Model: "pc-model",
+	})
+
+	new := s.brands.Model("canonical", "pc-model", map[string]interface{}{
+		"architecture":   "amd64",
+		"kernel":         "pc-kernel",
+		"gadget":         "pc",
+		"base":           "core18",
+		"store":          "switched-store",
+		"required-snaps": []interface{}{"new-required-snap-1", "new-required-snap-2"},
+		"revision":       "1",
+	})
+
+	freshStore := &freshSessionStore{}
+	testStore = freshStore
+
+	s.newFakeStore = func(devBE storecontext.DeviceBackend) snapstate.StoreService {
+		mod, err := devBE.Model()
+		c.Check(err, IsNil)
+		if err == nil {
```

Comment:
(nitpick) We check "c.Check(err, IsNil)" before the "if err == nil" - can we remove the `if` or do you prefer it because it gives cleaner output on test failures? I think I slightly prefer it it would go away because it increase the cognitive load slightly when reading this.